src/VecSim/algorithms/svs/svs_tiered.h

```diff
@@ -0,0 +1,726 @@
+            std::scoped_lock lock(this->flatIndexGuard, this->mainIndexGuard, this->journal_mutex);
+            ret = this->frontendIndex->addVector(blob, label);
+            ret = std::max(ret - svs_index->deleteVectors(&label, 1), 0);
+            journal.emplace_back(label, true);
+            index_update_needed = this->backendIndex->indexSize() > 0 ||
+                                  this->journal.size() >= this->updateJobThreshold;
+        }
+
+        if (index_update_needed) {
+            scheduleSVSIndexUpdate();
+        }
+
+        return ret;
+    }
+
+    int deleteVector(labelType label) override {
+        int ret = 0;
+        auto svs_index = GetSVSIndex();
+        if (this->getWriteMode() == VecSim_WriteInPlace) {
+            assert([&] {
+                std::shared_lock<std::shared_mutex> flat_lock(this->flatIndexGuard);
+                return !this->frontendIndex->isLabelExists(label);
+            }());
+
+            std::scoped_lock lock(this->updateJobMutex, this->mainIndexGuard);
+            return svs_index->deleteVectors(&label, 1);
+        }
+
+        bool label_exists = [&] {
+            std::shared_lock<std::shared_mutex> flat_lock(this->flatIndexGuard);
+            return this->frontendIndex->isLabelExists(label);
+        }();
+
+        if (label_exists) {
+            std::scoped_lock lock(this->flatIndexGuard, this->mainIndexGuard, this->journal_mutex);
+            if (this->frontendIndex->isLabelExists(label)) {
+                ret = this->frontendIndex->deleteVector(label);
+                assert(ret == 1 && "unexpected deleteVector result");
+                journal.emplace_back(label, false);
+            }
+            ret += svs_index->deleteVectors(&label, 1);
+            assert(ret < 2 && "deleteVector: vector duplication in both indices");
+        } else {
+            std::scoped_lock lock(this->mainIndexGuard);
+            ret += svs_index->deleteVectors(&label, 1);
+        }
+        return ret;
+    }
+
+    size_t indexSize() const override {
+        std::shared_lock<std::shared_mutex> flat_lock(this->flatIndexGuard);
+        std::shared_lock<std::shared_mutex> main_lock(this->mainIndexGuard);
+        return this->frontendIndex->indexSize() + this->backendIndex->indexSize();
+    }
+
+    size_t indexLabelCount() const override {
+        std::shared_lock<std::shared_mutex> flat_lock(this->flatIndexGuard);
+        std::shared_lock<std::shared_mutex> main_lock(this->mainIndexGuard);
+        return this->frontendIndex->indexLabelCount() + this->backendIndex->indexLabelCount();
+    }
+    size_t indexCapacity() const override {
+        std::shared_lock<std::shared_mutex> flat_lock(this->flatIndexGuard);
+        std::shared_lock<std::shared_mutex> main_lock(this->mainIndexGuard);
+        return this->frontendIndex->indexCapacity() + this->backendIndex->indexCapacity();
+    }
+
+    double getDistanceFrom_Unsafe(labelType label, const void *blob) const override {
```

Comment:
this function is called unsafe because it doesn't acquire any locks
how come it didn't lead to a deadlock in the tests? :thinking: